### DISCUSSION TITLE
2.2 and Data Augmentation

### DISCUSSION THREAD

**Discussion by caseybasichis** [explosion/spaCy#4739] [Help: Coding & Implementations] [answered]
I can’t find any documentation about the Data Augmentation pipeline mentioned on the spacy blog.  It was unclear if it was an announcement or an already included feature. 

I’m looking to add custom augmentation before fine tuning the spacy-transformers with a maskedLM task.

**Comment by adrianeboyd** [comment]
It's included, but there's no detailed documentation yet. The data augmentation is an option for `spacy train` and `GoldCorpus.train_docs()` right now. It generates variants of tokens (mostly punctuation for variation in ASCII vs. unicode quotes and hyphens) and lowercases some portion of sentences where variants are added. We need to experiment more with the values of the parameters here.

I'm planning to add a new extension to vary individual characters within tokens to handle things like `s` vs. `ș` vs. `ş` in Romanian.

**Comment by svlandeg** [comment, marked as answer]
Happy to report that we've significantly refactored this code for the upcoming [v3](https://nightly.spacy.io/usage/v3), which you can already test by installing the current Release Candidate as  [`spacy-nightly`](https://pypi.org/project/spacy-nightly/).

The [data augmentation](https://nightly.spacy.io/usage/training#data-augmentation) functionality is now provided as customizable options in the [config file](https://nightly.spacy.io/usage/training#config) used for [training](https://nightly.spacy.io/usage/training#config-custom) a new pipeline. You can use this data augmentation functionality easily from the provided [`spacy.Corpus.v1`](https://nightly.spacy.io/api/corpus#config) reader:

```
[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0

[corpora.train.augmenter]
@augmenters = "spacy.orth_variants.v1"
level = 0.1
lower = 0.5

[corpora.train.augmenter.orth_variants]@readers = "srsly.read_json.v1"
path = "corpus/orth_variants.json"
```

or you can use it in a custom implementation of your own - see also this related question: https://github.com/explosion/spaCy/discussions/6519.
